File: list_iteration/pl/mats_02.pl
#!/usr/bin/perl
use strict;
use warnings;

##########################################
## programmatic construction of 2D matrix 
## with list refs & push.
##
## author: vladimir kulyukin
##
## the camel always delivers...
##########################################

my @mat = ();

## add rows into 2d mat
my @row0 = (1..4);
push(@mat, \@row0);
my @row1 = (5..8);
push(@mat, \@row1);
my @row2 = (9..12);
push(@mat, \@row2);

## print each rwo in a for loop
sub print_2d_mat {
    my @mat = @_;
    for(my $r = 0; $r <= $#mat; $r++) {
	print "@{$mat[$r]}\n";
    }
}

## print with two for loops
sub print_2d_mat2 {
    my @mat = @{$_[0]};
    my $num_rows = $_[1];
    my $num_cols = $_[2];
    for(my $r = 0; $r < $num_rows; $r++) {
	for(my $c = 0; $c < $num_cols; $c++) {
	    print $mat[$r][$c], "\t";
	}
	print "\n";
    }
}

## test
print_2d_mat(@mat);
print_2d_mat2(\@mat, 3, 4);

File: list_iteration/py/mats_02.py
#!/usr/bin/python
import sys

####################################
## programmatic construction of 2D mats
##
## author: vladimir kulyukin
#####################################

row0 = range(1, 5)
row1 = range(5, 9)
row2 = range(9, 13)

mat = []
mat.append(row0)
mat.append(row1)
mat.append(row2)

def print_2d_mat(mat, num_rows, num_cols):
  for r in xrange(0, num_rows):
    for c in xrange(0, num_cols):
      print mat[r][c], '\t',
    print

def print_2d_mat2(mat, num_rows, num_cols):
  for r in xrange(0, num_rows):
    for c in xrange(0, num_cols):
      sys.stdout.write(str(mat[r][c])+'\t')
    sys.stdout.write('\n')

print_2d_mat(mat, 3, 4)
print_2d_mat2(mat, 3, 4)
